crates/runtime/src/dataconnector/graphql.rs

```diff
@@ -0,0 +1,285 @@
+     https://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+use arrow::array::RecordBatch;
+use arrow::datatypes::{Schema, SchemaRef};
+use arrow::error::ArrowError;
+use arrow_json::reader::infer_json_schema_from_iterator;
+use arrow_json::ReaderBuilder;
+use async_trait::async_trait;
+use datafusion::error::DataFusionError;
+use datafusion::execution::context::SessionState;
+use datafusion::logical_expr::Expr;
+use datafusion::physical_plan::ExecutionPlan;
+use itertools::Itertools;
+use reqwest::header::{HeaderMap, HeaderValue};
+use reqwest::header::{CONTENT_TYPE, USER_AGENT};
+use secrets::{get_secret_or_param, Secret};
+use serde_json::Value;
+use url::Url;
+
+use crate::component::dataset::Dataset;
+use datafusion::datasource::{MemTable, TableProvider, TableType};
+use snafu::prelude::*;
+use std::any::Any;
+use std::io::Cursor;
+use std::pin::Pin;
+use std::sync::Arc;
+use std::{collections::HashMap, future::Future};
+
+use super::{DataConnector, DataConnectorFactory};
+
+#[derive(Debug, Snafu)]
+pub enum Error {
+    #[snafu(display("Unable to handle request: {source}"))]
+    UnableToHandleRequest { source: reqwest::Error },
+
+    #[snafu(display("Failed to request data:\n {response}"))]
+    FailedToRequestData { response: String },
+
+    #[snafu(display("Unable to parse response: {source}"))]
+    UnableToParseResponse { source: reqwest::Error },
+
+    #[snafu(display("Unable to read batch: {source}"))]
+    UnableToReadBatch { source: ArrowError },
+
+    #[snafu(display("Unable to collect batch: {source}"))]
+    UnableToCollectBatch { source: ArrowError },
+
+    #[snafu(display("Unable to infer schema: {source}"))]
+    UnableToInferSchema { source: ArrowError },
+
+    #[snafu(display("Invalid object access: {message}"))]
+    InvalidObjectAccess { message: String },
+}
+
+pub type Result<T, E = Error> = std::result::Result<T, E>;
+
+pub struct GraphQL {
+    secret: Option<Secret>,
+    params: Arc<HashMap<String, String>>,
+}
+
+impl DataConnectorFactory for GraphQL {
+    fn create(
+        secret: Option<Secret>,
+        params: Arc<HashMap<String, String>>,
+    ) -> Pin<Box<dyn Future<Output = super::NewDataConnectorResult> + Send>> {
+        Box::pin(async move {
+            let graphql = Self { secret, params };
+            Ok(Arc::new(graphql) as Arc<dyn DataConnector>)
+        })
+    }
+}
+
+enum Auth {
+    Basic(String, Option<String>),
+    Bearer(String),
+}
+
+struct GraphQLClient {
+    client: reqwest::Client,
+    endpoint: Url,
+    query: String,
+    json_path: String,
+    auth: Option<Auth>,
+}
+
+impl GraphQLClient {
+    async fn execute(&self) -> Result<(Vec<Vec<RecordBatch>>, SchemaRef)> {
```

Comment:
Ideally this should return Sendable..Stream, as graphql supports paging, etc. but can be done in a separate PR.

Reply:
https://github.com/spiceai/spiceai/issues/1594